lib/vdsm/virt/thinp.py

```diff
@@ -225,18 +230,25 @@ def monitor_volumes(self):
         """
         Check and extend drives if needed.
         """
-        drives = self._monitored_volumes()
-        if not drives:
-            return
-
-        if not self._update_block_info(drives):
+        extend_timeout = config.getfloat("thinp", "extend_timeout")
+        if not self._lock.acquire(timeout=extend_timeout):
+            self._log.debug("Monitor is busy, retrying later")
             return
-
-        for drive in drives:
-            try:
-                self._extend_drive_if_needed(drive)
-            except ImprobableResizeRequestError:
-                break
+        try:
+            drives = self._monitored_volumes()
+            if not drives:
+                return
+
+            if not self._update_block_info(drives):
+                return
+
+            for drive in drives:
+                try:
+                    self._extend_drive_if_needed(drive)
+                except ImprobableResizeRequestError:
+                    break
+        finally:
+            self._lock.release()
```

Comment:
We can replace this lock with a lock on the drive around _extend_drive_if_needed().

Reply:
Great, I have a version replacing the monitor lock with a lock per drive. I move the locking patch to the end of the PR, and will post in 2 parts.